src/arch/wasm/CodeGen.zig

```diff
@@ -5022,19 +5036,33 @@ fn fptrunc(func: *CodeGen, operand: WValue, given: Type, wanted: Type) InnerErro
     const given_bits = given.floatBits(func.target);
     const wanted_bits = wanted.floatBits(func.target);
 
-    if (wanted_bits == 32 and given_bits == 64) {
-        try func.emitWValue(operand);
+    var current_operand = operand;
+    var current_bits = given_bits;
+    var current_ty = given;
+
+    if (wanted_bits == 32 and given_bits > 64) {
+        current_operand = try func.fptrunc(current_operand, given, Type.f64);
+        current_bits = 64;
+        current_ty = Type.f64;
+    }
+
+    if (wanted_bits == 32 or wanted_bits == 16 and given_bits == 64) {
+        try func.emitWValue(current_operand);
         try func.addTag(.f32_demote_f64);
-        return WValue{ .stack = {} };
+        current_operand = .stack;
+        current_bits = 32;
+        current_ty = Type.f32;
```

Comment:
Same as above, let's keep the original code (pre-PR) with just the `TODO` replaced with the dynamic compiler-rt call.